### DISCUSSION TITLE
Mikro-Orm exported types issue

### DISCUSSION THREAD

**Discussion by jeroni** [mikro-orm/mikro-orm#5457] [General]
I'm developing a Postgresql application and I have some doubts about exported types.

I was upgrading my code to V6 and reading the migration guide in this section:
https://mikro-orm.io/docs/upgrading-v5-to-v6#all-drivers-now-re-export-the-mikro-ormcore-package

I was happy to directly have a Sql EntityManager as I'm using Postgres and avoid the casting hell when I needed to create, for example, a QueryBuilder.

The problem I found is the mismatch of driver types, even importing from Postgresql implementation (I imagine is a direct forward from the core):

```ts
import { Entity, EntityManager, EventArgs, EventSubscriber, PrimaryKey, Property, t } from '@mikro-orm/postgresql';

@Entity({
    schema: 'invoicing',
})
class Customer {
    @PrimaryKey({type: t.bigint})
    id: string;

    @Property()
    name!: string;

    @Property()
    phone!: string;

    constructor(name: string, phone: string) {
        this.name = name;
        this.phone = phone;
    }
}

class ExampleSubscriber implements EventSubscriber<Customer> {
    getSubscribedEntities() {
        return [ Customer ];
    }

    beforeCreate(args: EventArgs<Customer>) {
        this.checkPhoneNumber(args.entity, args.em); // args.em is incompatible with Postgresql entity manager
    }

    beforeUpdate(args: EventArgs<Customer>) {
        this.checkPhoneNumber(args.entity, args.em); // args.em is incompatible with Postgresql entity manager
    }

    private checkPhoneNumber(customer: Customer, em: EntityManager) {
        // Being a Postgresql entity manager I have query builder available without casting to sql entity manager
        const qb = em.qb(Customer); // This is only for example purposes
    }
}
```
In this example the EventArgs, even being imported from postgresql do not forward the "correct" entity manager type, so I have two solutions as far as I can undestand:

1 - Cast the entity manager to sql entity manager when calling `this.checkPhoneNumber(args.entity, args.em as SqlEntityManager)`
2 - Use the core EntityManager, but I have to cast to SqlEntityManager again to create the query builder and have imports from the two parts: core and postgres.

I don't know if I'm missing something, but it would be good all the subtypes exported from the specific database implementation are already casted to Postgresql driver to be fully compatible.

**Comment by B4nan** [comment]
This is not v6 specific, `EventArgs` always referred to the EM from core. But thanks to the v6 changes, we could indeed add a new driver-specific child type for `EventArgs`. 

But it could mean a lot of duplication, will need to take a closer look, might not be worth it.

**Comment by jeroni** [comment]
I undestand it's a lot of work and there are other types to consider too, such as `RequestContext`.
My rationale was to minimize casting and prevent confusion among developers. Admittedly, navigating Mikro's strong typing can be challenging; I'm light years away from your level of comprehension!

**Reply by B4nan** [reply]
Btw you could always use type overrides in your project, as you know what driver you are using, the hard part is doing this dynamically with generics and whatnot.

**Reply by jeroni** [reply]
We've considered it, but it's not worth the pain. We think Mikro has a great workflow and those are minor issues, at least for us. The point is if it could be confusing for a new developer trying to undestand the imports and levels of abstraction, even with the solid docs.
I thought if it was a casting at the top level of the driver implementation would be easier, if it's not that easy I wouldn't bother about it.
